### DISCUSSION TITLE
Multiple fetches

### DISCUSSION THREAD

**Discussion by ryliecc** [neuefische/questions#634] [Web]
_english version below_

Hallo,
ich versuche momentan meine API anzubinden. Es gibt einige Endpoints, die sehr selten ihre Daten ändern, weswegen ich sie bei Appstart komplett cachen möchte um später bequem darauf zugreifen zu können. Ich benutze SWR zum fetchen.
Der aktuelle Stand ist gepusht, der branch heißt "display-api-data" und ist [hier](https://github.com/ryliecc/capstone_webdev_wanikani-app/tree/display-api-data) zu finden.
Einer dieser Endpoints ist `subjects`. Die API schickt immer 1000 Objekte auf einmal zurück. Ich hätte gerne, dass die App beim ersten Gebrauch von useSubjects direkt alle pages auf einmal fetcht (sprich mehrere fetches hintereinander ausführt) und alles cacht. Momentan verstehe ich aber nur wie ich eine custom hook für einen ganz speziellen endpoint, sprich nur 1000 Objekte erstelle. Wie schaffe ich es, dass solange gefetcht wird bis alle Objekte angekommen sind?

Die Daten ändern sich sehr selten, aber ich brauche sie ständig. Daher ist es für mich sinnvoller alles zu cachen und dann zu sortieren anstatt immer wieder mit verschiedenen endpoints (z.B. `subjects?types=vocabulary` oder `subjects?levels=8,42`) zu fetchen.

Danke für die Hilfe im vorraus!
x Rylie

---------

Hey,
I am currently trying to connect my API to my application. There are some endpoints which are not updated frequently, but I need their data all the time. That's why I want to cache them. I use SWR for that. I pushed the latest update, the branch is called "display-api-data" and you can find it [here](https://github.com/ryliecc/capstone_webdev_wanikani-app/tree/display-api-data).
One of those endpoints is `subjects`. The API sends back 1000 objects at a time, however there are way more subjects. That would mean I would have to fetch several times, each times using another endpoint for the specific subject I want (e.g. `subjects?types=vocabulary`). I would like to avoid that by fetching all pages of the subject endpoint when starting the app. However, until now I only understand how to build a custom hook for one single fetch. How can I make my SWR hook fetch several times until I received everything?
I would like to sort it after caching everything, so I can avoid making requests that only return 25 objects, because that is not very efficient. My API has a rate limit of 60 requests per hour, so I can't risk doing too many inefficient requests at once.

Thanks for the help in advance!
x Rylie

**Comment by julialoeschel** [comment]
Hey, 
the Link is not working, and could you provide  the link to the api-docs as well? 
thanks

**Reply by ryliecc** [reply]
That's because the branch does not exist anymore, I already merged it. You can find the main branch [here](https://github.com/ryliecc/capstone_webdev_wanikani-app). The docs for the API can be found [here](https://docs.api.wanikani.com/20170710/#introduction). Additionally if you wanna play around with it yourself you can use the personal API token `af559ce2-4a1b-4cf1-85a1-b2e9bcfc16ef`.

**Comment by marcbendixen** [comment]
Hey @ryliecc 👋 
if I understand your point correctly, you have the following options:

1. Everything remains as it currently is. You use the `useSubjects` hook with the parameters to request a specific information based on the `endpointPath`. The problem could be that each user makes a lot of requests to the API by switching the view/page. The good thing is that the data from the API is already filtered for you.
2. Another solution could be that you get all the data from the API and filter it yourself. But since you only get 1000 objects for a request (for example: `https://api.wanikani.com/v2/subjects`) you need to chain the requests. Inside the response you get the key `next_url`, for example `https://api.wanikani.com/v2/subjects?page_after_id=1000`. Making a fetch to this URL gives you the next `next_url` and so on. When there is no more `next_url` left, the value of this key is `null`. With this you can try to build a loop to fetch the informatione from all URLs and put the response data together in one variable or state. This data (with all objects, not only 1000) could then be filtered and used for the rest of the app. If a user navigates to another page, no more request to the API, just the filtered data from the initial fetch which should only be loaded once (for example in the `_app.js`)

I hope this helps you a bit to figure it out.
